[Diagnostics] Add a variant of the extraneous property wrapper unwrap diagnostic for function arguments

sl opened the pull request:
Adds an instance of the missing property wrapper unwrap diagnostic for function arguments. This pull request depends on https://github.com/apple/swift/pull/25507. 

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->
Related to [SR-10928](https://bugs.swift.org/browse/SR-10928).

sl committed:
Sema: Add a diagnostic for inserting a $ to remove extraneous property wrapper unwrap

sl committed:
Sema: Add a diagnostic for missing property wrapper unwrap operations in function calls

sl committed:
Test: Add a test for the new extraneous property wrapper unwrap diagnostic

sl committed:
Sema: Remove handling for property wrapper composition in diagnostic

sl committed:
Test: Add a test for missing property wrapper unwrap operators in function calls

xedin commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -2175,6 +2175,66 @@ bool ConstraintSystem::repairFailures(
+    }
+
+    Expr *argAnchor = nullptr;
+    if (auto *AE = dyn_cast<ApplyExpr>(loc->getAnchor())) {
```
Easier way to do this is to use - `simplifyLocatorToAnchor` - for apply argument path it would return argument you are looking for :)

sl replied:
Ahh, glad to know that's a thing!

xedin commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -2175,6 +2175,66 @@ bool ConstraintSystem::repairFailures(
+      break;
+
+    // Find the resolved overload related to this
+    auto resolvedOverload = getResolvedOverloadSets();
```
Two suggestions here:

 - please extract logic related to finding something in resolvedOverloadSets into a method on ConstraintSystem e.g. `findSelectedOverloadFor(Expr *)`
- This whole block could be extracted into a separate function which returns a wrapper so it could be opened and processed here.

sl replied:
I'll implement that method in the primary PR: https://github.com/apple/swift/pull/25507 and migrate this one to make use of it!

xedin replied:
I'd really like to get this into 5.1 piecemeal, so it'd be great to do it here and then rebase with the primary PR which might not get into 5.1.

sl replied:
Ah, I think there's a bit of confusion. This pull request is a superset of the other one. If function arguments are the priority, I can remove the member part from this pull request though and then finish building this one out though.

xedin replied:
Ah okay, I'm just lost in the PRs sorry - cool then, let's do it there and rebase this one.

xedin commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -4774,6 +4834,46 @@ ConstraintSystem::SolutionKind ConstraintSystem::simplifyMemberConstraint(
+        resolvedOverload = resolvedOverload->Previous;
+      }
+
+      if (resolvedOverload && resolvedOverload->Choice.isDecl()) {
```
See my previous comment, if the logic related to finding wrapper is extracted into its own function it could be reused here.

xedin reviewed (changes requested):
Looks good, just needs some more work!

harlanhaskins commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -992,6 +992,9 @@ ERROR(missing_address_of,none,
       "yielding mutable value of type %0 requires explicit '&'",
       (Type))
+ERROR(missing_property_wrapper_unwrap,none,
+      "wrapped value of type %0 must be unwraped to value of type %1",
```
Terminology here: are we going to consider the wrapped value as "wrapped" when it's used without the `$`, and then "unwrapped" when it's used with the `$`?

Semantically, adding the $ stops the property from being unwrapped.

Also small typo: `unwraped` -> `unwrapped`

sl replied:
Good point, I did get this backwards. That said, I can’t think of a good way to talk about this the other way around. Saying something like: “the property wrapper must not be unwrapped” sounds very strange because all property wrappers are implicitly unwrapped so the user didn’t do anything to unwrap it. It gets even weirder for people new to the feature when the suggested fix is to add a $ to *prevent* the implicit unwrapping. Any idea for what the diagnostic for this should be?

xedin replied:
Thanks Harlan! I actually haven't looked at the message... It looks like it should be `extraneous_property_unwrap` and message should get adjusted accordingly e.g.  `extraneous property %0 unwrap to value of type %1, use `$` to refer to wrapper type.` where `$0` is `DeclName` and $1 is `Type`. @harlanhaskins WDYT?

xedin replied:
or `cannot convert type %0 to expected property wrapper type %1, use $ to refer to wrapper type.`

harlanhaskins replied:
Hm, I know we have precedent for `extraneous` but I think we can simplify this text. Maybe `property %0 will be unwrapped to value of type %1, use '$' to refer to the property's wrapper`

harlanhaskins replied:
Or maybe `use '$' to refer to the @%2 property wrapper`, where `%2` is the `Type` of the wrapper?

xedin replied:
+1

xedin commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -4774,6 +4834,46 @@ ConstraintSystem::SolutionKind ConstraintSystem::simplifyMemberConstraint(
+          if (decl->hasAttachedPropertyWrapper()) {
+            auto rawWrapperTy = decl->getAttachedPropertyWrapperType(0);
+            auto wrapperTy =
+                openUnboundGenericType(rawWrapperTy, resolvedOverload->Locator);
```
Here is one more suggestion - instead of opening wrapper type here and then adding constraint to match it with value type, what you can do is form `BoundGenericType` from wrapper type and pass already known value type as it's generic argument.

xedin commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -4774,6 +4834,46 @@ ConstraintSystem::SolutionKind ConstraintSystem::simplifyMemberConstraint(
+                wrappedInfo.valueVar->getValueInterfaceType());
+            if (!valueTy->hasError()) {
+              addConstraint(ConstraintKind::Equal, valueTy, baseTy,
+                            resolvedOverload->Locator);
```
I think we don't need this constraint if you follow my previous suggestion.

sl committed:
Sema: Change the diagnostic message for extraneous property wrapper unwraps

Additionally refactor some of the repeatable logic for getting useful
property wrapper related types out into methods on ConstraintSystem.

sl commented:
I addressed a the review feedback for this PR over in https://github.com/apple/swift/pull/25507 which I plan to get merged first. Those changes will be reflected here once that gets merged.

sl committed:
WIP on missing-dollar-sign-diagnostics: a90b634f99 Sema: Change the diagnostic message for extraneous property wrapper unwraps

sl committed:
index on missing-dollar-sign-diagnostics: a90b634f99 Sema: Change the diagnostic message for extraneous property wrapper unwraps

sl committed:
Sema: Change property extraneous property wrapper diagnostic to use getPropertyWrapperBackingPropertyType

sl committed:
Test: Add test for diagnostic when property wrappers have multiple generic arguments

sl committed:
Sema: Update extraneous property wrapped unwrap diagnostic for func args to use new methods

Changes the extraneous property wrapper unwrap diagnostic for function arguments to
use the new utility methods for working with property wrappers defined on
ConstraintSystem.

sl commented:
Work on this was moved to #25889.
